### ISSUE TITLE
~compile only watches some files

### ISSUE THREAD

**Issue by blast-hardcheese** [sbt/sbt#3695]
### Setup

```
mkdir -p project src/main/scala/foo
echo sbt.version=1.0.3 > project/build.properties
echo 'println("woo")' > src/main/scala/App.scala
echo 'println("woo")' > src/main/scala/foo/bar.scala
```

### Launch sbt `~compile`

```
$ sbt
sbt> ~compile
```

### Change files

1. Changes to `src/main/scala/App.scala` always trigger a recompile
1. Changes to `src/main/scala/foo/bar.scala` never trigger a recompile
1. Create a new file, `src/main/scala/foo/baz.scala`. Changes to `baz.scala` trigger a recompile until the next full sbt launch (close and open again).

## problem

`~` doesn't always track files in directories that should be covered by the file watcher

## expectation

`~` always tracks files in directories that are covered by the file watcher

## notes

sbt version: 1.0.3

**Comment by blast-hardcheese** [comment]
@dwijnand This is the issue you requested of me from Gitter, please let me know if I can be of any help going forward

**Comment by blast-hardcheese** [comment]
Possibly related to #3687

**Comment by dwijnand** [comment]
Thanks @blast-hardcheese.

**Comment by OlegYch** [comment]
for me `~` doesn't see changes in tests, but when i modify main source `~testOnly` runs twice
same for `~compile` though

**Comment by dwijnand** [comment]
I can't reproduce this. I followed the given steps and it all works as intended on my box:

```scala
> ~compile
[info] Compiling 2 Scala sources to /s/t-watches/target/scala-2.12/classes ...
[error] /s/t-watches/src/main/scala/App.scala:1:1: expected class or object definition
[error] println("woo")
[error] ^
[error] /s/t-watches/src/main/scala/foo/bar.scala:1:1: expected class or object definition
[error] println("woo")
[error] ^
[error] two errors found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 2 s, completed 16-Nov-2017 16:05:04
1. Waiting for source changes... (press enter to interrupt)
[info] Compiling 2 Scala sources to /s/t-watches/target/scala-2.12/classes ...
[error] /s/t-watches/src/main/scala/App.scala:1:1: expected class or object definition
[error] println("I've changed App.scala")
[error] ^
[error] /s/t-watches/src/main/scala/foo/bar.scala:1:1: expected class or object definition
[error] println("woo")
[error] ^
[error] two errors found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 0 s, completed 16-Nov-2017 16:05:19
2. Waiting for source changes... (press enter to interrupt)
[info] Compiling 2 Scala sources to /s/t-watches/target/scala-2.12/classes ...
[error] /s/t-watches/src/main/scala/App.scala:1:1: expected class or object definition
[error] println("I've changed App.scala")
[error] ^
[error] /s/t-watches/src/main/scala/foo/bar.scala:1:1: expected class or object definition
[error] println("I've changed foo/bar.scala")
[error] ^
[error] two errors found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 1 s, completed 16-Nov-2017 16:05:30
3. Waiting for source changes... (press enter to interrupt)
[info] Compiling 2 Scala sources to /s/t-watches/target/scala-2.12/classes ...
[error] /s/t-watches/src/main/scala/App.scala:1:1: expected class or object definition
[error] println("I've changed App.scala")
[error] ^
[error] /s/t-watches/src/main/scala/foo/bar.scala:1:1: expected class or object definition
[error] println("I've changed foo/bar.scala (again)")
[error] ^
[error] two errors found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 0 s, completed 16-Nov-2017 16:05:36
4. Waiting for source changes... (press enter to interrupt)
[info] Compiling 3 Scala sources to /s/t-watches/target/scala-2.12/classes ...
[error] /s/t-watches/src/main/scala/App.scala:1:1: expected class or object definition
[error] println("I've changed App.scala")
[error] ^
[error] /s/t-watches/src/main/scala/foo/bar.scala:1:1: expected class or object definition
[error] println("I've changed foo/bar.scala (again)")
[error] ^
[error] /s/t-watches/src/main/scala/foo/baz.scala:1:1: expected class or object definition
[error] println("I've created foo/baz.scala")
[error] ^
[error] three errors found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 0 s, completed 16-Nov-2017 16:05:55
5. Waiting for source changes... (press enter to interrupt)
[info] Compiling 3 Scala sources to /s/t-watches/target/scala-2.12/classes ...
[error] /s/t-watches/src/main/scala/App.scala:1:1: expected class or object definition
[error] println("I've changed App.scala")
[error] ^
[error] /s/t-watches/src/main/scala/foo/bar.scala:1:1: expected class or object definition
[error] println("I've changed foo/bar.scala (again)")
[error] ^
[error] /s/t-watches/src/main/scala/foo/baz.scala:1:1: expected class or object definition
[error] println("I've changed foo/baz.scala")
[error] ^
[error] three errors found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 0 s, completed 16-Nov-2017 16:06:04
6. Waiting for source changes... (press enter to interrupt)
```

Going to switch to #3687, hopefully I can reproduce that.

**Comment by blast-hardcheese** [comment]
@dwijnand I was just similarly unable to reproduce this on a linux VM, despite being able to reproduce it exactly the same way on my Macbook Air.

Upgrading Java now to see if that helps on my mac
**Update**: Upgrading to java version `1.8.0_151` did not impact reproducibility on OSX

### Mac specs
#### OS version
10.12.6 (16G1036)
#### Java version
java version "1.8.0_92"
Java(TM) SE Runtime Environment (build 1.8.0_92-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.92-b14, mixed mode)
#### SBT launcher
Experiment 1. Standard, 1.0.0 from Homebrew
Experiment 2. paulp's launcher

### CentOS specs
#### Java version
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
#### SBT launcher
paulp's launcher

**Comment by blast-hardcheese** [comment]
I was able to reproduce on a completely different mac, so I'm wondering if this is an OSX issue. That computer was also running 10.12.6

**Comment by blast-hardcheese** [comment]
I bisected and ended up on b9a4d3291. `-Dsbt.watch.mode=nio` is definitely slow, although it seems to behave as expected, both in this test example and the project I was initially having issues with

**Comment by blast-hardcheese** [comment]
I wonder, @dwijnand, can you repro with `-Dsbt.watch.mode=polling`? I've got a script to repro, switching between `nio` and `polling` gives consistent results: `nio` always works as expected, `polling` always fails.

**Comment by DylanArnold** [comment]
For what it's worth I tried to repro this issue on Ubuntu linux a week ago to try and find out if it was the same issue I was having with incremental compilation in the `1.0.x` series.

I was unable to reproduce the problem using the original steps.

**Comment by eatkins** [comment]
I added an sbt plugin that resolves this issue for me on my mac: https://github.com/swoval/MacOSXWatchService

**Comment by dwijnand** [comment]
I expect this to be fixed with https://github.com/sbt/io/issues/86 too.

**Comment by dwijnand** [comment]
Re-opening to track sbt/io needs releasing for inclusion in 1.1.0-RC2.

**Comment by eed3si9n** [comment]
New io with the fix was merged in #3816
